### PR TITLE
feat: update mocked endpoints

### PR DESCRIPTION
**Pull request by andrewelamb**
fixes [fds-688](https://sagebionetworks.jira.com/browse/FDS-688)

- Changed path of mock endpoint from storages/projects/{@projectId}/datasets to /assetTypes/{assetType}/assetViews/{assetViewId}/projects/{projectId}/datasets
- Changed path of endpoint from storage/asset-views/{assetView}/projects/{projectId}/manifests to assetTypes/@{assetType}/assetViews/@{assetViewId}/projects/@{projectId}/manifests.yaml
- Added asset type as a parameter in the two endpoints above and organized storage resources under asset types
- Updated assetView to assetViewId
- Changed definition of datasetId and manifest (see Dataset.yaml and Manifest.yaml)
- Added mypy,  pylint
- Updated black, schematicpy, flack, and click versions
- Added tests for handing errors 401, 403, 500
- lint command now does mypy/pylint/black checking
- lint-fix command now does black fixing
- `handle_exceptions` function is used as a decorator to handle various error conditions and can be used my multiple endpoint functions
- `get_asset_storage_class` function is used to determine the asset class based on the `asset_type` argument. This will be useful in the future when there are multiple asset types to choose from
- `get_project_datasets` and `get_project_manifests` are function used to interact with the storage class. The purpose of these functions is to make it easier to mock synapse functionality for testing.
- removed config handler function

### CODE REVIEW

**Comment by tschaffter** [comment]
@andrewelamb There is one check that fails and needs to be fixed before the PR can be reviewed.

**Comment by tschaffter** [comment]
Unit and integration tests should mock calls to external services like Synapse, thus eliminating the need for the CI workflow to depend on secrets.

**Comment by tschaffter** [comment]
@andrewelamb Congrats on fixing the tests of your first Sage Monorepo PR! 🎉 

I see that the task that build images failed because of OC projects. As mentioned above, updating this feature branch with the content of `main` should remove the OC projects from the list of `affected` projects.

**Comment by tschaffter** [comment]
The CI workflow is running out of storage space because of the numerous affected projects.

```console
...
    [creator]     Adding label 'io.buildpacks.build.metadata'
    [creator]     Adding label 'io.buildpacks.project.metadata'
    [creator]     Adding label 'org.springframework.boot.version'
    [creator]     Setting default process type 'web'
    [creator]     Saving ghcr.io/sage-bionetworks/openchallenges-challenge-service-base:local...
    [creator]     ERROR: failed to export: saving image: fetching base layers: write /tmp/imgutil.local.image.2749701756/22bbaf1fa50dced77ae49a3be9119a74802577036986acf544fc44047f0abbe8/layer.tar: no space left on device
```

@andrewelamb This is a general issue that I will investigate over the week. Meanwhile, updating this feature branch with `main` should remove all the OC projects from the list of affected projects. The smaller number of affected projects will also speedup the execution of the workflow.

**Review by tschaffter — APPROVED** [review]
LGTM

**Comment by andrewelamb** [comment]
@linglp @milen-sage Did you want to take a look before I merge?

**Review by milen-sage — APPROVED** [review]
Thanks @andrewelamb ! The PR provides a good overview of the overall structure, and that looks good to me. I'll defer to @linglp on reviewing the REST API logic in more detail, but that can happen later once its merged, too. 

I'm assuming this PR is large because is building up the codebase and in the future PRs would be scoped to smaller piece of code?

**Comment by andrewelamb** [comment]
@milen-sage to a degree. The schema is spread across many files, so if you add an endpoint you will likely have to add/change several files. On the other hand, much of this is automatically generated so those wouldn't have to be checked during a PR.

**Review by linglp — CHANGES_REQUESTED** [review]
I think this PR should really be split into two or even three. For a big PR like this, I would appreciate if there's a more detailed description of what gets changed and what gets added. Here's a summary that I think could be added to PR description: 

## changelog
* Changed path of mock endpoint from `storages/projects/{@projectId}/datasets` to  `/assetTypes/{assetType}/assetViews/{assetViewId}/projects/{projectId}/datasets`
* Added mock implementation of `/assetTypes/{assetType}/assetViews/{assetViewId}/projects/{projectId}/datasets`
* Changed path of endpoint from `storage/asset-views/{assetView}/projects/{projectId}/manifests` to `assetTypes/@{assetType}/assetViews/@{assetViewId}/projects/@{projectId}/manifests.yaml`
* Added asset type as a parameter in the two endpoints above and organized storage resources under asset types 
* Updated assetView to assetViewId
* Changed definition of datasetId and manifest (see Dataset.yaml and Manifest.yaml) 
* Added mypy, black, pylint
* Added tests for handing erros 401, 403, 500

File: apps/schematic/api/schematic_api/controllers/storage_controller_impl.py
Comment (andrewelamb) [reply apps/schematic/api/schematic_api/controllers/storage_controller_impl.py:102]: when you do dict.get("key") if they key isn't in the dict, None is returned.

File: libs/schematic/api-description/src/components/schemas/Manifest.yaml
Comment (andrewelamb) [reply libs/schematic/api-description/src/components/schemas/Manifest.yaml:16]: The current storage/projects/manifests endpoint, which this replaces, returns those things so I'm returning them here as well.

File: apps/schematic/api/schematic_api/controllers/storage_controller_impl.py
Comment (andrewelamb) [reply apps/schematic/api/schematic_api/controllers/storage_controller_impl.py:142]: Yes, you are correct I'm fixing this and adding a test to catch this.

File: libs/schematic/api-description/src/components/schemas/Manifest.yaml
Comment (andrewelamb) [reply libs/schematic/api-description/src/components/schemas/Manifest.yaml:16]: Oh also schemtic db  uses storage/projects/manifests, and makes use of the component name.

File: apps/schematic/api/schematic_api/controllers/storage_controller_impl.py
Comment (linglp) [reply apps/schematic/api/schematic_api/controllers/storage_controller_impl.py:102]: Oh I see it now. Thanks.

File: libs/schematic/api-description/src/components/schemas/Manifest.yaml
Comment (linglp) [reply libs/schematic/api-description/src/components/schemas/Manifest.yaml:16]: I just tried using `storage/project/manifest` endpoint locally, and it returned: 
```
[
  [
    [
      "syn30988361",
      "TestFolder"
    ],
    [
      "syn30988380",
      "synapse_storage_manifest (2).csv"
    ],
    [
      "Biospecimen",
      "Biospecimen"
    ]
  ]
]
```
I think based on this output, it returned manifest id, manifest name, and also component name? But it doesn't return `datasetId` and `datasetName`?

Comment (andrewelamb) [reply libs/schematic/api-description/src/components/schemas/Manifest.yaml:16]: "syn30988361",
 "TestFolder"

Those are the dataset id and name

Comment (andrewelamb) [reply libs/schematic/api-description/src/components/schemas/Manifest.yaml:16]: Yeah, here's what the schematic function returns:

```
 Returns: A list of datasets per project; metadata manifest Synapse ID for each dataset; and the corresponding schema component of the manifest
                 as a list of tuples, one for each manifest:
                    [
                        (
                            (datasetId, dataName),
                            (manifestId, manifestName),
                            (componentSchemaLabel, componentSchemaLabel) TODO: # get component name from schema
                        ),
                        ...
                    ]
```

Comment (linglp) [reply libs/schematic/api-description/src/components/schemas/Manifest.yaml:16]: Oh I see it now. Thanks for explaining.

**Comment by linglp** [comment]
@andrewelamb I tried running your test locally, and it seems like that it is not working: 
I did: `poetry run pytest test_storage_controller_endpoints.py::TestDatasets::test_success`, and here's the output: 
```
    self.assertEqual(response.status_code, status_code, message)
E   AssertionError: 500 != 200 : Response body is : {
E     "detail": "[Errno 2] No such file or directory: '/workspaces/sage-monorepo/apps/schematic/api/schematic/api/schematic_api/config.yml'",
E     "status": 500,
E     "title": "Internal error"
E   }
```
And also, when I ran test in `test_storage_controller_impl.py`, tests were also failing with 500 errors.

**Comment by andrewelamb** [comment]
@linglp 
`No such file or directory: '/workspaces/sage-monorepo/apps/schematic/api/schematic/api/schematic_api/config.yml'`
Does this file not exist in your directory?
[config.yml](https://github.com/Sage-Bionetworks/sage-monorepo/blob/main/apps/schematic/api/schematic_api/config.yml)

**Comment by linglp** [comment]
@andrewelamb The `config.yml` is there but the test is pointing at the wrong location. I think to run the test, you have to first: `cd /workspaces/sage-monorepo/apps/schematic/api`, and then `poetry run pytest schematic_api/test/test_storage_controller_endpoints.py::TestDatasets`.. This probably has something to do with how `config_handler` was written: 
```
def config_handler(asset_view: Optional[str] = None) -> None:
    """Load config file and update asset view if needed
    Args:
        asset_view (str): asset view
    """
    CONFIG.load_config(
        "../../schematic/api/schematic_api/config.yml", asset_view=asset_view
    )
```
If you change the relative path to an absolute path, I think it could avoid the error.

File: apps/schematic/api/schematic_api/controllers/storage_controller_impl.py
Comment (linglp) [reply apps/schematic/api/schematic_api/controllers/storage_controller_impl.py:119]: Thanks for sharing the design perspective. I understand it better now.. You could keep the code this way.

**Comment by andrewelamb** [comment]
@linglp That config handler is going to get removed when we update to the next schematic version, as the config will be optional at that point. Since the tests are passing in CI/CD is it ok if we leave it for now?

**Comment by andrewelamb** [comment]
@linglp Could you check this now? I updated Schematic to the latest version, which allowed for removal of the problematic config_handler function.

**Comment by linglp** [comment]
@andrewelamb sorry for the late reply. I just got my dev container working again. When I executed `http://localhost:7081/api/v1/assetTypes/synapse/assetViews/syn23643253/projects/syn26251192/manifests` endpoint, I got an error message: 
```
{
  "detail": "The 'DATA' configuration field was accessed, but it hasn't been set yet, presumably because the schematic.CONFIG.load_config() method hasn't been run yet. Alternatively, you can re-run this code with the 'SCHEMATIC_CONFIG' environment variable set to the config.yml file, which will be automatically loaded.",
  "status": 500,
  "title": "Internal error"
}
```
I think it has something to do with not having `config.yml` in schematic. 

For simplicity, I think you could revert back to previous version of schematic and update `CONFIG.load_config` file path. After I approve, you could merge the changes you have now and create a separate PR for updating to the latest version of schematic.

**Comment by andrewelamb** [comment]
@linglp I'm, pretty sure that's an error message that was removed during the configuration.py refactor in Schematic. I'm not seeing either of those errors on my local api. My guess is there's somethign wrong with your local setup, where it isn't using the latest schematic version.

**Comment by andrewelamb** [comment]
@linglp

**Comment by andrewelamb** [comment]
@linglp You can run `nx run prepare-python` to make sure your python packages are up to date

**Review by linglp — APPROVED** [review]
Approve it for now.

**Review by tschaffter — APPROVED** [review]

**Pull request merged**
